### DISCUSSION TITLE
Initialize context based on user parameters + Context privacy

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#2453] [Ideas] [answered]
Hi again. I would like to request a new feature that enables machine users to pass parameters to a function (i.e. context initializater) whose return value will be the actual initial context.

This should also allow proper spawning of actors within the context of the parent machine during context initialization. Currently, if `spawn` is called from outside the machine, actors are not properly linked to the parent machine.

Another benefit to this feature is context privacy enforced by the limited allowed parameters to be set by the user during context initialization (i.e. not all context fields are assignable by the user during initialization).

I think initialization logic sometimes belongs to the machine itself and not to the user.
Initialization parameters here are similar to the concept of `props` in frameworks like Vue and React.

**Comment by davidkpiano** [comment, marked as answer]
One way this can already be achieved is with a machine factory:

```js
const createUserMachine = (id) => createMachine({
  context: () => ({
    // initial spawn works
    clientRef: spawn(createClientMachine(id)),
    userId: id
  }),
  // ...
});
```

**Reply by geekox86** [reply]
This is nice. Thx!

**Reply by geekox86** [reply]
@davidkpiano may I request to add the option to pass a function to `machine.withContext` so that we do not need to recreate the whole machine with every call to the factory (i.e. reuse the machine)?

```
const machine = createMachine({...});
const createMachine = (id) => machine.withContext(() => { child: spawn(createChildMachine(id)) });
```

**Reply by Andarist** [reply]
This would still recreate the whole machine:
https://github.com/statelyai/xstate/blob/10944fa6195a3146c7c120d7c3c5c5defeff4e72/packages/core/src/StateNode.ts#L485-L489

Only the input arguments are being "cached" on the `this` but no real reuse is happening between the original machine and its "fork".

**Reply by davidkpiano** [reply]
Do you think it would be worthwhile to structurally share as much as we can?

**Reply by ChrisShank** [reply]
I find it helpful as long as the internals can be understood in some amount of isolation!

**Reply by Andarist** [reply]
> Do you think it would be worthwhile to structurally share as much as we can?

That would probably be a preferable solution but also not something of the utmost importance 😉
